iOS/Flows/Conversation/ConversationListCollectionViewLayout.swift

```diff
@@ -40,11 +50,76 @@ class ConversationListCollectionViewLayout: UICollectionViewFlowLayout {
                                                    right: (collectionView.width - itemWidth).half)
 
         // NOTE: Subtracting 1 to ensure there's enough vertical space for the cells.
-        let verticalSpacing = collectionView.height - itemHeight - 1
+        let verticalSpacing = clamp(collectionView.height - itemHeight - 1, min: 0)
         self.sectionInset = UIEdgeInsets(top: 0,
                                          left: 0,
                                          bottom: verticalSpacing,
                                          right: 0)
         self.minimumLineSpacing = Theme.ContentOffset.standard.value
     }
+
+    private func sendCenterUpdateEventIfNeeded(withContentOffset contentOffset: CGPoint) {
+        var cid: ConversationId?
+        if let centeredItem = self.getCenteredItem(forContentOffset: contentOffset) {
+            cid = self.delegate?.conversationListCollectionViewLayout(self,
+                                                                      cidFor: centeredItem.indexPath)
+        }
+        if self.previousCenteredCID != cid {
+            self.previousCenteredCID = cid
+            self.delegate?.conversationListCollectionViewLayout(self, didUpdateCentered: cid)
+        }
+    }
+
+    // MARK: - Scrolling Behavior
+
+    override func targetContentOffset(forProposedContentOffset proposedContentOffset: CGPoint,
+                                      withScrollingVelocity velocity: CGPoint) -> CGPoint {
+
+        guard let collectionView = self.collectionView else { return .zero }
+
+        guard let centeredItem = self.getCenteredItem(forContentOffset: proposedContentOffset) else {
+            return super.targetContentOffset(forProposedContentOffset: proposedContentOffset,
+                                             withScrollingVelocity: velocity)
+        }
+
+        // Always scroll so that a cell is centered when we stop scrolling.
+        let newXOffset = centeredItem.frame.centerX - collectionView.halfWidth
+
+        return CGPoint(x: newXOffset, y: proposedContentOffset.y)
+    }
+
+    override func targetContentOffset(forProposedContentOffset proposedContentOffset: CGPoint) -> CGPoint {
+        self.sendCenterUpdateEventIfNeeded(withContentOffset: proposedContentOffset)
```

Comment:
This fixes a bug. If an item is inserted, it may change what is centered without changing the content offset.

This function is called whenever an insert/delete happens.